feat: support configurable secrets hiding in UI and API

okurz opened the pull request:
Motivation:
Some test settings might contain sensitive information that doesn't follow
the default redaction pattern (^_SECRET_ or _PASSWORD). Users should be able
to specify custom patterns for redaction to prevent leaking secrets in logs,
test settings files, and the web interface.

Design Choices:
- Added _HIDE_SECRETS_REGEX job setting to allow custom redaction patterns.
- Optimized regex compilation in OpenQA::Log::redact_settings.
- Added redacted_settings_hash to OpenQA::Schema::Result::Jobs for centralized
  redaction.
- Updated to_hash in Jobs result class to support optional redaction.
- Updated Web UI templates and API controllers to use redacted settings.

Benefits:
- Improved security by preventing accidental leakage of custom secrets.
- Greater flexibility for users to protect sensitive data.
- Consistent redaction across logs, worker files, web UI, and API.

Related progress issue: https://progress.opensuse.org/issues/162086

okurz commented on lib/OpenQA/Log.pm:
```diff
@@ -208,8 +208,11 @@ sub setup_log ($app, $logfile = undef, $logdir = undef, $level = undef) {
 sub redact_settings ($vars) {
-    return {map { $_ !~ qr/(^_SECRET_|_PASSWORD)/ ? ($_ => $vars->{$_}) : ($_ => '[redacted]') } keys %$vars};
+    my $hide_re = '^_SECRET_|_PASSWORD';
+    $hide_re .= "|$vars->{_HIDE_MATCH_RE}" if $vars->{_HIDE_MATCH_RE};
```
Both suggestions covered including the fallback to ignore the additional custom stuff and still apply the default. I am still doing a concatenation to form the final regex. With more procedural code we could also use `if /$default_re/ || /$custom_re/` but IMHO it wouldn't be better. WDYT?

okurz replied:
done

Martchus commented on lib/OpenQA/Log.pm:
```diff
@@ -208,8 +208,13 @@ sub setup_log ($app, $logfile = undef, $logdir = undef, $level = undef) {
+    my $hide_re = '^_SECRET_|_PASSWORD';
+    my $custom_hide_re = eval { qr/$vars->{_HIDE_SECRETS_REGEX}/ } if $vars->{_HIDE_SECRETS_REGEX};
+    $vars->{_HIDE_SECRETS_REGEX} = "(invalid regex specified: $@)" if $@;
+    $hide_re .= "|$custom_hide_re" if $custom_hide_re;
+    return {map { /($hide_re)/ ? ($_ => '[redacted]') : ($_ => $vars->{$_}) } keys %$vars};
```
It would probably make sense to compile `$hide_re` into a regex only once outside the `map` call:

```suggestion
    $hide_re = $custom_hide_re ? qr/($hide_re|$custom_hide_re)/ : qr/($hide_re)/;
    return {map { ($_ =~ $hide_re) ? ($_ => '[redacted]') : ($_ => $vars->{$_}) } keys %$vars};
```

okurz committed:
feat: support configurable secrets hiding in UI and API

Motivation:
Some test settings might contain sensitive information that doesn't follow
the default redaction pattern (^_SECRET_ or _PASSWORD). Users should be able
to specify custom patterns for redaction to prevent leaking secrets in logs,
test settings files, and the web interface.

Design Choices:
- Added _HIDE_SECRETS_REGEX job setting to allow custom redaction patterns.
- Optimized regex compilation in OpenQA::Log::redact_settings.
- Added redacted_settings_hash to OpenQA::Schema::Result::Jobs for centralized
  redaction.
- Updated to_hash in Jobs result class to support optional redaction.
- Updated Web UI templates and API controllers to use redacted settings.

Benefits:
- Improved security by preventing accidental leakage of custom secrets.
- Greater flexibility for users to protect sensitive data.
- Consistent redaction across logs, worker files, web UI, and API.

Related progress issue: https://progress.opensuse.org/issues/162086

okurz commented:
all suggestions incorporated and covered redacting in UI+API

Martchus commented on lib/OpenQA/Schema/Result/Jobs.pm:
```diff
@@ -432,6 +432,11 @@ sub settings_hash ($self, $prefetched = undef) {
 }
 
+sub redacted_settings_hash ($self, $prefetched = undef) {
+    require OpenQA::Log;
```
Why can't we just use `OpenQA::Log` normally here?

okurz replied:
yes, we can. Seems like an old left-over for localized testing. fixed
